File: backend/src/utils/spotify.rs
use base64::engine::{general_purpose, Engine as _};
use dotenvy::var;
use reqwest::header::{HeaderMap, CONTENT_TYPE};
use serde_json::Value;
use std::collections::HashMap;

const URL: &str = "https://api.spotify.com/v1/playlists/6onj45lQm8DiLFT8nQQCOj/tracks";

async fn get_bearer() -> String {
    let (secret, client) = (var("SPOTIFY_SECRET").unwrap(), var("SPOTIFY_CLIENT").unwrap());
    let b64 = general_purpose::STANDARD.encode(format!("{}:{}", client, secret));
    let client = reqwest::Client::new();

    let params = HashMap::from([("grant_type", "client_credentials")]);
    let mut headers = HeaderMap::new();
    headers.insert("Application", format!("Basic: {}", b64).parse().unwrap());

    let res = client.post(URL).form(&params).headers(headers).send().await;
    let clean_response: String = match res {
        Ok(v) => v.text().await.unwrap(),
        Err(_) => return String::from("")
    };
    let json: Value = serde_json::from_str(&clean_response).unwrap();
    let field_value = json["bearer"].as_str().unwrap().to_string();
    field_value
}

File: backend/utils/spotify.py

from os import getenv as ge
import httpx
import time
from utils.database import Database, EXPIRE
from typing import List


class Track:
    name: str
    artists: str
    url: str

    def __init__(self, name, artists, url):
        self.name = name
        self.artists = artists
        self.url = url


class SpotifyToken:
    bearer = None
    expire = 0

    @staticmethod
    async def cache(k: str, tracks: List[Track]):
        dict_tracks = []
        for track in tracks:
            dict_tracks.append(vars(track))
        await Database.redisConn.set(k, str(dict_tracks), EXPIRE)

    @staticmethod
    async def genToken():
        SpotifyToken.expire = time.time() + 3600
        SPOTIFY_CLIENT, SPOTIFY_SECRET = ge(
            "SPOTIFY_CLIENT"), ge("SPOTIFY_SECRET")
        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://accounts.spotify.com/api/token",
                headers={"Content-Type": "application/x-www-form-urlencoded"},
                data={"grant_type": "client_credentials",
                      "client_id": SPOTIFY_CLIENT, "client_secret": SPOTIFY_SECRET}
            )
            response_json = response.json()
            SpotifyToken.bearer = response_json["access_token"]

    @staticmethod
    async def getTracks(items: int) -> List[Track]:
        if time.time() >= SpotifyToken.expire:
            await SpotifyToken.genToken()
        async with httpx.AsyncClient() as client:
            redis_data = await Database.redisConn.get("tracks")
            if redis_data:
                converted_list = eval(redis_data)
                return converted_list[-items:]
            response = await client.get("https://api.spotify.com/v1/playlists/6onj45lQm8DiLFT8nQQCOj/tracks", headers={"Authorization": f"Bearer {SpotifyToken.bearer}"})
            response_json = response.json()
            tracks = []
            for track in response_json["items"]:
                track_indexed = track["track"]
                artists = ""
                for artist in track_indexed["artists"]:
                    artists += f"{artist['name']}, "
                artists = artists[:-2]
                tracks.append(
                    Track(track_indexed["name"], artists, track_indexed["external_urls"]["spotify"]))
            await SpotifyToken.cache("tracks", tracks[-items:])
            return tracks[-items:]
